jobs: add RevertFailed terminal state for jobs that fail during revert

adityamaru opened the pull request:
Previously, a failure during revert would mark the job as "failed" and
indicate in the error string that manual cleanup is required. This job
was however gc'eable if the retention time has elapsed.  This would
leave no trace of the job which is bad, since we expect users/support to
cleanup the partial work done by the job.

This change adds a new terminal state `RevertFailed` that does not allow
the job to get gc'ed.

Fixes: #59542

Release note: None

adityamaru committed:
jobs: add RevertFailed terminal state for jobs that fail during revert

Previously, a failure during revert would mark the job as "failed" and
indicate in the error string that manual cleanup is required. This job
was however gc'eable if the retention time has elapsed.  This would
leave no trace of the job which is bad, since we expect users/support to
cleanup the partial work done by the job.

This change adds a new terminal state `RevertFailed` that does not allow
the job to get gc'ed.

Fixes: #59542

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63311)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice

> Reviewed 5 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63311#-:-MY80F8_A9MpVYWt7uyQ:bb76hxo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pbardea)



<!-- Sent from Reviewable.io -->

pbardea commented on pkg/jobs/jobs.go:
```diff
@@ -172,7 +176,7 @@ func deprecatedIsOldSchemaChangeJob(payload *jobspb.Payload) bool {
 // after which the job should never be updated again.
 func (s Status) Terminal() bool {
-	return s == StatusFailed || s == StatusSucceeded || s == StatusCanceled
+	return s == StatusFailed || s == StatusSucceeded || s == StatusCanceled || s == StatusRevertFailed
```
Given this job state is treated as `Terminal`, should they also be considered a candidate for GC?

ajwerner replied:
A major motivation here is to not GC these things.

pbardea replied:
Ah, I had incorrectly thought that we check if a job is Terminal to determine if it's a candidate for GC, but that's not the case. Sorry for the noise.

adityamaru commented:
TFTRs!

bors r=ajwerner,pbardea

adityamaru commented:
Oops @pbardea I just realized you hadn't LGTM'ed this. If you have any comments I'll address them separately!

adityamaru commented:
rebased on master

adityamaru commented:
bors r+

adityamaru commented:
Dropped off my radar. Updated the test to not check for the orphaned jobs being cleaned up because we no longer have this logic after - https://github.com/cockroachdb/cockroach/pull/63730.

adityamaru commented:
bors r+
